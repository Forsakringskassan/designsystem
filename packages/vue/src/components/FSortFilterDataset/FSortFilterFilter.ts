import { isSet } from "@fkui/logic";

function includesAllSearchTerms(
    item: Record<string, string | number | undefined>,
    filterAttributes: string[],
    searchTerms: string[],
): boolean {
    const values = filterAttributes
        .map((it) => {
            const value = item[it];
            return isSet(value)
                ? value.toString().toLocaleLowerCase()
                : undefined;
        })
        .filter(Boolean);

    for (const searchTerm of searchTerms) {
        const match = values.find((it) => it?.includes(searchTerm));
        if (!match) {
            return false;
        }
    }
    return true;
}

export function filter<T>(
    list: T[],
    filterAttributes: string[],
    searchString: string,
): T[] {
    searchString = searchString.trim();
    if (searchString.trim() === "") {
        return list;
    }

    const searchTerms = searchString
        .split(/\s+/)
        .map((word) => word.toLocaleLowerCase());

    return list.filter((item) =>
        includesAllSearchTerms(
            item as unknown as Record<string, string | number | undefined>,
            filterAttributes,
            searchTerms,
        ),
    );
}
