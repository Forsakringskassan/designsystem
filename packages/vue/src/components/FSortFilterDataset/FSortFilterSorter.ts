import { isSet } from "@fkui/logic";

export function sort<T>(
    list: T[],
    sortAttribute: string,
    ascending: boolean,
): T[] {
    return list.sort((item1, item2) =>
        compare(item1, item2, sortAttribute as keyof T, ascending),
    );
}

function compare<T>(
    item1: T,
    item2: T,
    attribute: keyof T,
    ascending: boolean,
): number {
    const value1 = item1[attribute];
    const value2 = item2[attribute];

    if (!isSet(value1) || !isSet(value2)) {
        // Null values are always unsorted at the end of the list
        return nullCompare(value1, value2);
    }
    if (typeof value1 === "string" && typeof value2 === "string") {
        return fixOrder(value1.localeCompare(value2), ascending);
    }
    if (typeof value1 === "number" && typeof value2 === "number") {
        return fixOrder(numberCompare(value1, value2), ascending);
    }
    if (typeof value1 === "boolean" && typeof value2 === "boolean") {
        return fixOrder(booleanCompare(value1, value2), ascending);
    }

    if (typeof value1 !== typeof value2) {
        throw Error(
            `Sorting is not supported for rows containing different types (attribute '${attribute.toString()}')`,
        );
    }

    throw Error(
        `Sorting is only supported for types number, string and boolean (attribute '${attribute.toString()}')`,
    );
}

// This is used instead of list.reverse to keep null values
// at the end of the list and in unsorted order
function fixOrder(order: number, ascending: boolean): number {
    return ascending ? order : order - order * 2;
}

function booleanCompare(value1: boolean, value2: boolean): number {
    if (value1 === value2) {
        return 0;
    } else if (value1 > value2) {
        return 1;
    }
    return -1;
}

function numberCompare(value1: number, value2: number): number {
    return value1 - value2;
}

function nullCompare(value1: unknown, value2: unknown): number {
    if (!isSet(value1) && !isSet(value2)) {
        return 0;
    } else if (!isSet(value1)) {
        return 1;
    }
    return -1;
}
