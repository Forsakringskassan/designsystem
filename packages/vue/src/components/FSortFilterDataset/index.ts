export {
    type FSortFilterDatasetInterface,
    type FSortFilterDatasetMountCallback,
    type FSortFilterDatasetSortCallback,
    FSortFilterDatasetInjected,
} from "./FSortFilterDatasetInterface";
export { default as FSortFilterDataset } from "./FSortFilterDataset.vue";
export { type SortOrder } from "./sort-order";
