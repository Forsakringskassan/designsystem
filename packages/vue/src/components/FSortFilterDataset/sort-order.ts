/**
 * Represents the attributes used when sorting.
 *
 * @public
 */
export interface SortOrder {
    attribute: string;
    name: string;
    ascendingName: string;
    ascending: boolean;
    id: number;
}
