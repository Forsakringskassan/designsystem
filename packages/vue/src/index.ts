/* include typescript declarations for vite static asset handling (e.g. `?raw`) */
/// <reference types="vite/client" />

export * from "./components";
export {
    type UseSlotUtils,
    type WithoutInstance,
    useSlotUtils,
} from "./composables";
export {
    config,
    type FKUIConfig,
    FKUIConfigButtonOrder,
    type FKUIContext,
    type MaybeWithFKUIContext,
    setRunningContext,
} from "./config";
export * from "./internal-components";
export * from "./plugins";
export {
    type ComponentUnmountEvent,
    type ComponentValidityEvent,
    type ErrorItem,
    type GroupValidityEvent,
    type ListArray,
    type ListItem,
    type UnknownItem,
    ErrorData,
    ErrorViewData,
    FormErrorList,
    MenuAction,
} from "./types";
export * from "./utils";
