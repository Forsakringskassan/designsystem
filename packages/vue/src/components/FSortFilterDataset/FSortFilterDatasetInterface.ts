import { inject } from "vue";

/**
 * @public
 * @param attribute - is the field that got sorted
 * @param ascending - is the order the fields got sorted
 */
export type FSortFilterDatasetSortCallback = (
    attribute: string,
    ascending: boolean,
) => void;

/**
 * @public
 */
export type FSortFilterDatasetMountCallback = (columnNames: string[]) => void;

/**
 * This interface can be implemented (provided) by Vue components
 * supporting filtering. Not required.
 *
 * @public
 */
export interface FSortFilterDatasetInterface {
    /**
     * Sort the dataset
     * @param attribute - is the field that gets sorted
     * @param ascending - is the order the fields get sorted
     */
    sort(attribute: string, ascending: boolean): void;

    /**
     * Callback function that is called when dataset get sorted
     */
    registerCallbackOnSort(callback: FSortFilterDatasetSortCallback): void;

    /**
     * Called when FSortFilterDataset is mounted.
     * Returns the sortable field names.
     */
    registerCallbackOnMount(callback: FSortFilterDatasetMountCallback): void;
}

/**
 * Empty dummy methods to be used when no FSortFilterDataset is used
 *
 * @public
 */
export function FSortFilterDatasetInjected(): FSortFilterDatasetInterface {
    return {
        sort: inject("sort", () => undefined) as (
            attribute: string,
            ascending: boolean,
        ) => void,
        registerCallbackOnSort: inject(
            "registerCallbackOnSort",
            () => undefined,
        ) as (callback: FSortFilterDatasetSortCallback) => void,
        registerCallbackOnMount: inject(
            "registerCallbackOnMount",
            () => undefined,
        ) as (callback: FSortFilterDatasetMountCallback) => void,
    };
}
