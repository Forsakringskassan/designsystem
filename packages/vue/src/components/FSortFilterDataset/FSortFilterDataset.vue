<script setup lang="ts" generic="T">
import { type PropType, type Ref, computed, nextTick, onMounted, provide, ref, useTemplateRef, watch } from "vue";
import { focus, debounce, alertScreenReader, TranslationService } from "@fkui/logic";
import { IFlex, IFlexItem } from "../../internal-components/IFlex";
import { FSelectField } from "../FSelectField";
import { FIcon } from "../FIcon";
import { FTextField } from "../FTextField";
import { useTranslate } from "../../plugins";
import { FSortFilterDatasetSortCallback, FSortFilterDatasetMountCallback } from "./FSortFilterDatasetInterface";
import { sort } from "./FSortFilterSorter";
import { filter } from "./FSortFilterFilter";
import { type SortOrder } from "./sort-order";

const $t = useTranslate();

const searchString = ref("");
const defaultSortValue = { attribute: "", name: "", ascendingName: "", ascending: false, id: 0 };
const sortAttribute = ref<SortOrder>(defaultSortValue);
const sortFilterResult = ref<T[]>([]) as Ref<T[]>;
const debouncedFilterResultset = debounce(filterResultset, 250);

let tableCallbackOnSort: FSortFilterDatasetSortCallback = () => {
    /* do nothing */
};

let tableCallbackSortableColumns: FSortFilterDatasetMountCallback = () => {
    /* do nothing */
};

const props = defineProps({
    /**
     * The data that you wish to sort or filter.
     */
    data: {
        type: Array as PropType<T[]>,
        required: true,
        default: () => [],
    },
    /**
     * All the attributes you want to enable sorting for and the corresponding name to display in the dropdown.
     * Structured as `{attributeName: "Name for dropdown", secondAttributeName: "Name for dropdown"}`
     */
    sortableAttributes: {
        type: Object as PropType<Record<string, string>>,
        required: true,
    },
    /**
     * If set the data will be sorted by this attribute by default.
     */
    defaultSortAttribute: {
        type: String,
        required: false,
        default: () => "",
    },
    /**
     * Show/hides the sort dropdown.
     */
    showSort: {
        type: Boolean,
        required: false,
        default: true,
    },
    /**
     * Show/hides the filter input.
     */
    showFilter: {
        type: Boolean,
        required: false,
        default: true,
    },
    /**
     * Set placeholder text in filter input field.
     * Default is textkey "fkui.sort-filter-dataset.placeholder.filter"
     */
    placeholderFilter: {
        type: String,
        required: false,
        default: TranslationService.provider.translate("fkui.sort-filter-dataset.placeholder.filter", "Sök"),
    },
    /**
     * The order the data will be sorted by if defaultSortAttribute has been set.
     */
    defaultSortAscending: {
        type: Boolean,
        required: false,
        default: () => true,
    },
});

const emit = defineEmits<{
    /**
     * Emitted when the data is sorted.
     *
     * @arg items - The sorted data.
     */
    datasetSorted: [items: T[]];

    /**
     * Emits the used sorting attributes.
     *
     * @arg sortAttribute - The attributes used when sorting the data.
     */
    usedSortAttributes: [sortAttribute: SortOrder];
}>();

const showClearButton = computed(() => {
    return searchString.value.length > 0;
});

const sortOrders = computed((): SortOrder[] => {
    const arr = [] as SortOrder[];
    let id = 0;
    Object.keys(props.sortableAttributes).forEach((key: string) => {
        arr.push({
            attribute: key,
            name: props.sortableAttributes[key],
            ascendingName: $t("fkui.sort-filter-dataset.label.ascending", "stigande"),
            ascending: true,
            id: id++,
        });
        arr.push({
            attribute: key,
            name: props.sortableAttributes[key],
            ascendingName: $t("fkui.sort-filter-dataset.label.descending", "fallande"),
            ascending: false,
            id: id++,
        });
    });
    return arr;
});

const filterAttributes = computed(() => {
    return Object.keys(props.sortableAttributes);
});

provide("sort", (attribute: string, ascending: boolean) => {
    const foundAttribute = sortOrders.value.find((item) => {
        return item.attribute === attribute && item.ascending === ascending;
    });

    if (foundAttribute) {
        sortAttribute.value = foundAttribute;
    } else {
        sortAttribute.value = { attribute: "", ascending: false } as SortOrder;
    }

    sortFilterData();

    emit("usedSortAttributes", sortAttribute.value);
});

provide("registerCallbackOnSort", (callback: FSortFilterDatasetSortCallback) => {
    tableCallbackOnSort = callback;
});

provide("registerCallbackOnMount", (callback: FSortFilterDatasetMountCallback) => {
    tableCallbackSortableColumns = callback;
});

onMounted(() => {
    tableCallbackSortableColumns(Object.keys(props.sortableAttributes));
});

watch(
    () => props.data,
    () => {
        if (props.defaultSortAttribute !== "") {
            const foundAttribute = sortOrders.value.find((item) => {
                return item.attribute === props.defaultSortAttribute && item.ascending === props.defaultSortAscending;
            });
            if (foundAttribute) {
                sortAttribute.value = foundAttribute;
            }
        }
        sortFilterData();
    },
    { immediate: true, deep: true },
);

function sortFilterData(): void {
    const filteredData = filter(props.data, filterAttributes.value, searchString.value);

    if (sortAttribute.value.attribute === "") {
        sortFilterResult.value = filteredData;
    } else {
        sortFilterResult.value = sort([...filteredData], sortAttribute.value.attribute, sortAttribute.value.ascending);
    }
    // Await slot mount, otherwise if defaultSortAttribute is used the slot doesn't have time to register tableCallbackOnSort before this is called.
    nextTick(() => {
        tableCallbackOnSort(sortAttribute.value.attribute, sortAttribute.value.ascending);
    });
    emit("datasetSorted", sortFilterResult.value);
}

function onChangeSortAttribute(): void {
    sortFilterData();
    emit("usedSortAttributes", sortAttribute.value);
}

function onSearchInput(event: InputEvent): void {
    searchString.value = (event.target as HTMLInputElement).value;
    debouncedFilterResultset();
}

function onClickClearSearch(): void {
    searchString.value = "";
    sortFilterData();
    const input = useTemplateRef<HTMLElement>("search-field").value;
    focus(input);
}

function filterResultset(): void {
    sortFilterData();
    if (searchString.value === "") {
        alertScreenReader($t("fkui.sort-filter-dataset.aria-live.empty", "Sök redigera Sök tom"));
    } else {
        const searchAriaLive = $t(
            "fkui.sort-filter-dataset.aria-live.search",
            `Din sökning på "{{ search }}" gav {{ result }} träffar.`,
            {
                result: sortFilterResult.value.length,
                search: searchString.value,
            },
        );

        alertScreenReader(searchAriaLive);
    }
}
</script>

<template>
    <div class="sort-filter-dataset">
        <i-flex collapse gap="3x" wrap>
            <i-flex-item shrink align="center">
                <!--
                     @slot Slot for header to the left of the filter dropdown. Slot content is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ slotClass }"`.
                     @binding {string[]} slotClass - CSS classes to use for the SortFilter header element. Use with `:class="slotClass"`.
                -->
                <slot name="header" v-bind="{ slotClass: 'sort-filter-dataset__toolbar__header' }"></slot>
            </i-flex-item>

            <i-flex-item grow>
                <i-flex collapse float="right">
                    <i-flex-item v-if="showFilter" shrink align="center">
                        <div class="sort-filter-dataset__search">
                            <f-icon name="search" class="sort-filter-dataset__search__magnify-icon" />

                            <f-text-field
                                ref="search-field"
                                v-model="searchString"
                                inline
                                :placeholder="placeholderFilter"
                                maxlength="64"
                                @input="onSearchInput"
                            >
                                <span class="sr-only">{{ placeholderFilter }}</span>
                            </f-text-field>

                            <button
                                v-if="showClearButton"
                                type="button"
                                class="button button--discrete sort-filter-dataset__search__close-icon"
                                :title="$t('fkui.sort-filter-dataset.clear.filter', 'Rensa sökfält')"
                                @click="onClickClearSearch"
                            >
                                <f-icon name="close" />
                                <span class="sr-only">{{
                                    $t("fkui.sort-filter-dataset.clear.filter", "Rensa sökfält")
                                }}</span>
                            </button>
                        </div>
                    </i-flex-item>

                    <i-flex-item v-if="showSort" shrink align="center">
                        <f-select-field
                            v-model="sortAttribute"
                            class="sort-filter-dataset__sort"
                            inline
                            @change="onChangeSortAttribute"
                        >
                            <template #label>{{
                                $t("fkui.sort-filter-dataset.label.sort", "Sortera\u00A0på")
                            }}</template>

                            <option :value="defaultSortValue">
                                {{ $t("fkui.sort-filter-dataset.label.unsorted", "Välj") }}
                            </option>

                            <option v-for="sortOrder in sortOrders" :key="sortOrder.id" :value="sortOrder">
                                {{ sortOrder.name }} ({{ sortOrder.ascendingName }})
                            </option>
                        </f-select-field>
                    </i-flex-item>
                </i-flex>
            </i-flex-item>
        </i-flex>
        <!--
             @slot Slot for displaying the data. The data is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ sortFilterResult }"`.
             @binding {T[]} sortFilterResult - The sorted or filtered data.
        -->
        <slot v-bind="{ sortFilterResult }"></slot>
    </div>
</template>
