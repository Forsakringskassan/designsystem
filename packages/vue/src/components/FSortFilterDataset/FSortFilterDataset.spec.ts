import "html-validate/jest";
import { VueWrapper, mount } from "@vue/test-utils";
import { defineComponent } from "vue";
import {
    FSortFilterDatasetInjected,
    FSortFilterDatasetInterface,
} from "./FSortFilterDatasetInterface";
import FSortFilterDataset from "./FSortFilterDataset.vue";
import { type SortOrder } from "./sort-order";

const DATA = [
    { a: 1, b: "รถ", c: "dummy", d: {} },
    { a: 3, b: "a", c: "dummy", d: {} },
    { a: 2, b: "k", c: "dummy", d: {} },
];

const CASE_SENSITIVE_DATA = [
    { a: "Ab", b: 1 },
    { a: "ab", b: 1 },
    { a: "AA", b: 1 },
    { a: "aa", b: 1 },
];

const ATTRIBUTES = {
    a: "Column A",
    b: "Column B",
};

function createWrapper({ slots = {} } = {}): VueWrapper {
    return mount(FSortFilterDataset, {
        props: {
            data: DATA,
            sortableAttributes: ATTRIBUTES,
        },
        slots: { ...slots },
        global: {
            stubs: ["f-text-field", "f-icon"],
        },
    });
}

interface UnderlyingComponentData {
    sortableAttributes: string[];
    currentSortAttribute: string;
    currentAscending: boolean;
}

const UnderlyingComponent = defineComponent({
    data(): UnderlyingComponentData {
        return {
            sortableAttributes: [],
            currentSortAttribute: "",
            currentAscending: false,
        };
    },
    mounted() {
        this.registerCallbackOnSort(this.callbackOnSort);
        this.registerCallbackOnMount(this.callbackSortableColumns);
    },
    setup(): FSortFilterDatasetInterface {
        return FSortFilterDatasetInjected();
    },
    methods: {
        testSort(): void {
            this.sort("b", false);
        },
        triggersort(column: string, ascending: boolean): void {
            this.sort(column, ascending);
        },
        callbackSortableColumns(columns: string[]): void {
            this.sortableAttributes = columns;
        },
        callbackOnSort(columnName: string, ascending: boolean): void {
            this.currentSortAttribute = columnName;
            this.currentAscending = ascending;
        },
    },
    template: /* HTML */ `
        <div>
            <table>
                <tr v-for="item in data" :key="item.a">
                    <td>{{ item.a }}</td>
                    <td>{{ item.b }}</td>
                </tr>
            </table>
        </div>
    `,
    props: {
        data: {
            type: Array,
            required: true,
        },
    },
});

function createTableWrapper(
    testData: Array<Record<string, unknown>> = DATA,
): VueWrapper {
    const testComponent = {
        components: { UnderlyingComponent, FSortFilterDataset },
        template: /* HTML */ `
            <f-sort-filter-dataset
                :data="data"
                :sortable-attributes="attributes"
                default-sort-attribute="a"
                :default-sort-ascending="true"
            >
                <template #default="{ sortFilterResult }">
                    <underlying-component
                        :data="sortFilterResult"
                        id="underlying"
                    >
                    </underlying-component>
                </template>
            </f-sort-filter-dataset>
        `,
        data() {
            return {
                data: testData,
                attributes: ATTRIBUTES,
            };
        },
    };
    return mount(testComponent);
}

const tableTestComponentWithOutput = {
    components: { UnderlyingComponent, FSortFilterDataset },
    template: /* HTML */ `
        <f-sort-filter-dataset
            :data="data"
            :sortable-attributes="attributes"
            default-sort-attribute="a"
            :default-sort-ascending="true"
            @used-sort-attributes="usedSortAttributes($event)"
        >
            <template #default="{ sortFilterResult }">
                <underlying-component :data="sortFilterResult" id="underlying">
                </underlying-component>
            </template>
        </f-sort-filter-dataset>
    `,
    data() {
        return {
            data: [],
            attributes: ATTRIBUTES,
        };
    },
    methods: {
        usedSortAttributes(_sortAttributes: SortOrder) {
            /* intentionally empty, used for jest.spyOn */
        },
    },
};

function getUnderlyingComponent(
    wrapper: VueWrapper,
): InstanceType<typeof UnderlyingComponent> {
    return wrapper.getComponent<typeof UnderlyingComponent>("#underlying").vm;
}

describe("snapshots", () => {
    it("should match snapshot with a dropdown with attributes", () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });
});

it("should sort by default values", async () => {
    const wrapper = createTableWrapper();
    await wrapper.vm.$nextTick();
    const rows = wrapper.findAll("tr");

    expect(rows[0].find("td").text()).toBe("1");
    expect(rows[1].find("td").text()).toBe("2");
    expect(rows[2].find("td").text()).toBe("3");
});

it("should emit event with used attributes when sorting using table heading", async () => {
    const spyGetUsedSortAttributes = jest.spyOn(
        tableTestComponentWithOutput.methods,
        "usedSortAttributes",
    );

    const wrapper = mount(tableTestComponentWithOutput);

    getUnderlyingComponent(wrapper).triggersort("b", true);
    await wrapper.vm.$nextTick();

    expect(spyGetUsedSortAttributes).toHaveBeenCalledWith({
        attribute: "b",
        name: "Column B",
        ascendingName: "stigande",
        ascending: true,
        id: 2,
    });

    getUnderlyingComponent(wrapper).triggersort("b", false);

    expect(spyGetUsedSortAttributes).toHaveBeenCalledWith({
        attribute: "b",
        name: "Column B",
        ascendingName: "fallande",
        ascending: false,
        id: 3,
    });

    getUnderlyingComponent(wrapper).triggersort("", true);

    expect(spyGetUsedSortAttributes).toHaveBeenCalledWith({
        attribute: "",
        ascending: false,
    });
});

it("should sort by default values even when data changes", async () => {
    const wrapper = createTableWrapper();

    const newData = [
        { a: 7, b: "o" },
        { a: 5, b: "p" },
        { a: 6, b: "q" },
    ];
    await wrapper.setData({ data: newData });
    await wrapper.vm.$nextTick();

    const rows = wrapper.findAll("tr");
    expect(rows[0].find("td").text()).toBe("5");
    expect(rows[1].find("td").text()).toBe("6");
    expect(rows[2].find("td").text()).toBe("7");
});

it("should give an underlying component access to sort", async () => {
    const wrapper = createTableWrapper();

    getUnderlyingComponent(wrapper).testSort();
    await wrapper.vm.$nextTick();

    const rows = wrapper.findAll("tr");
    expect(rows[0].findAll("td")[1].text()).toBe("รถ");
    expect(rows[1].findAll("td")[1].text()).toBe("k");
    expect(rows[2].findAll("td")[1].text()).toBe("a");
});

it("should call the underlying component with available sortable atributes", () => {
    const wrapper = createTableWrapper();
    const sortableAttributes =
        getUnderlyingComponent(wrapper).sortableAttributes;
    expect(sortableAttributes).toEqual(["a", "b"]);
});

it("should call the underlying component when sorting changes", async () => {
    const wrapper = createTableWrapper();

    getUnderlyingComponent(wrapper).testSort();

    await wrapper.vm.$nextTick();
    const currentSortAttribute =
        getUnderlyingComponent(wrapper).currentSortAttribute;
    const currentAscending = getUnderlyingComponent(wrapper).currentAscending;
    expect(currentSortAttribute).toBe("b");
    expect(currentAscending).toBe(false);
});

it("should sort data when dropdown is changed", async () => {
    const wrapper = createTableWrapper();
    const options = wrapper.find("select").findAll("option");

    // Sort by Column B (Stigande)
    await options[3].setValue();
    let rows = wrapper.findAll("tr");
    expect(rows[0].findAll("td")[1].text()).toBe("a");
    expect(rows[1].findAll("td")[1].text()).toBe("k");
    expect(rows[2].findAll("td")[1].text()).toBe("รถ");

    // Sort by Column A (Stigande)
    await options[1].setValue();
    rows = wrapper.findAll("tr");
    expect(rows[0].findAll("td")[0].text()).toBe("1");
    expect(rows[1].findAll("td")[0].text()).toBe("2");
    expect(rows[2].findAll("td")[0].text()).toBe("3");
});

it("should emit event when dataset is sorted", async () => {
    const wrapper = createWrapper();
    await wrapper.vm.$nextTick();

    expect(wrapper.emitted()["datasetSorted"]).toBeTruthy();
    expect(wrapper.emitted()["datasetSorted"]).toEqual([[DATA]]);

    // Sort by Column B (Stigande)
    const options = wrapper.find("select").findAll("option");
    await options[3].setValue();
    expect(wrapper.emitted()["datasetSorted"]).toEqual([
        [DATA],
        [
            [
                { a: 3, b: "a", c: "dummy", d: {} },
                { a: 2, b: "k", c: "dummy", d: {} },
                { a: 1, b: "รถ", c: "dummy", d: {} },
            ],
        ],
    ]);
});

it("should emit event with used attributes when sorting using dropdown", async () => {
    const wrapper = createWrapper();
    await wrapper.vm.$nextTick();

    const options = wrapper.find("select").findAll("option");
    await options[3].setValue();
    expect(wrapper.emitted()["usedSortAttributes"][0]).toEqual([
        {
            attribute: "b",
            name: "Column B",
            ascendingName: "stigande",
            ascending: true,
            id: 2,
        },
    ]);

    await options[4].setValue();
    expect(wrapper.emitted()["usedSortAttributes"][1]).toEqual([
        {
            attribute: "b",
            name: "Column B",
            ascendingName: "fallande",
            ascending: false,
            id: 3,
        },
    ]);

    await options[0].setValue();
    expect(wrapper.emitted()["usedSortAttributes"][2]).toEqual([
        {
            attribute: "",
            ascending: false,
            ascendingName: "",
            id: 0,
            name: "",
        },
    ]);
});

it("should throw error when sorting objects", () => {
    const mountInvalidSort = (): void => {
        mount(FSortFilterDataset, {
            props: {
                data: DATA,
                sortableAttributes: {
                    d: "Column with objects",
                },
                defaultSortAttribute: "d",
            },
        });
    };
    expect(mountInvalidSort).toThrowErrorMatchingInlineSnapshot(`
        "Sorting is only supported for types number, string and boolean.
                    Attribute 'd' comparsion of types 'object' and 'object' is not supported."
    `);
});

it("should sort strings independent of case", async () => {
    const wrapper = createTableWrapper(CASE_SENSITIVE_DATA);
    const options = wrapper.find("select").findAll("option");
    // Sort by Column A (Stigande)
    await options[1].setValue();
    /* eslint-disable-next-line jest/no-large-snapshots -- easier to test with snapshot */
    expect(wrapper.get("table")).toMatchInlineSnapshot(`
        <table>
          <tr>
            <td>aa</td>
            <td>1</td>
          </tr>
          <tr>
            <td>AA</td>
            <td>1</td>
          </tr>
          <tr>
            <td>ab</td>
            <td>1</td>
          </tr>
          <tr>
            <td>Ab</td>
            <td>1</td>
          </tr>
        </table>
    `);
    // Sort by Column A (Fallande)
    await options[2].setValue();
    /* eslint-disable-next-line jest/no-large-snapshots -- easier to test with snapshot  */
    expect(wrapper.get("table")).toMatchInlineSnapshot(`
        <table>
          <tr>
            <td>Ab</td>
            <td>1</td>
          </tr>
          <tr>
            <td>ab</td>
            <td>1</td>
          </tr>
          <tr>
            <td>AA</td>
            <td>1</td>
          </tr>
          <tr>
            <td>aa</td>
            <td>1</td>
          </tr>
        </table>
    `);
});

describe("html-validate", () => {
    it("should require data attribute", () => {
        expect(
            "<f-sort-filter-dataset></f-sort-filter-dataset>",
        ).not.toHTMLValidate({
            message:
                '<f-sort-filter-dataset> is missing required "data" attribute',
        });
    });

    it("should require sortable-attributes attribute", () => {
        expect(
            "<f-sort-filter-dataset></f-sort-filter-dataset>",
        ).not.toHTMLValidate({
            message:
                '<f-sort-filter-dataset> is missing required "sortable-attributes" attribute',
        });
    });

    it("should require default slot", () => {
        expect(
            "<f-sort-filter-dataset></f-sort-filter-dataset>",
        ).not.toHTMLValidate({
            message:
                '<f-sort-filter-dataset> component requires slot "default" to be implemented',
        });
    });

    it("html should be valid", () => {
        expect(
            '<f-sort-filter-dataset data="test" sortable-attributes="test igen"><template #default></template></f-sort-filter-dataset>',
        ).toHTMLValidate();
    });
});
