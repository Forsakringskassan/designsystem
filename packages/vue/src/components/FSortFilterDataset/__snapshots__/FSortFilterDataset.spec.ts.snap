// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with a dropdown with attributes 1`] = `
<div
  class="sort-filter-dataset"
>
  <div
    class="iflex iflex--collapse iflex--gap-3x iflex--wrap"
  >
    
    <div
      class="iflex__item iflex--align-center iflex--shrink"
    >
      
      
    </div>
    <div
      class="iflex__item iflex--align-top iflex--grow"
    >
      
      <div
        class="iflex iflex--collapse iflex--float-right"
      >
        
        <div
          class="iflex__item iflex--align-center iflex--shrink"
        >
          
          <div
            class="sort-filter-dataset__search"
          >
            <f-icon-stub
              class="sort-filter-dataset__search__magnify-icon"
              library="f"
              name="search"
            />
            <f-text-field-stub
              disabled="false"
              id="fkui-vue-element-0001"
              inline="true"
              inputwidth="sm-12"
              labelwidth="sm-12"
              maxlength="64"
              modelvalue=""
              placeholder="Sök"
              type="text"
            />
            <!--v-if-->
          </div>
          
        </div>
        <div
          class="iflex__item iflex--align-center iflex--shrink"
        >
          
          <div
            class="select-field select-field--inline"
          >
            <div
              class=""
            >
              <label
                class="label"
                for="fkui-vue-element-0002"
              >
                
                
                Sortera på
                
                
                
                
                <!--v-if-->
              </label>
            </div>
            <div
              class="select-field__icon-wrapper"
            >
              <select
                class="select-field__select sort-filter-dataset__sort"
                id="fkui-vue-element-0002"
              >
                
                <option
                  value="[object Object]"
                >
                  Välj
                </option>
                
                <option
                  value="[object Object]"
                >
                  Column A (stigande) 
                </option>
                <option
                  value="[object Object]"
                >
                  Column A (fallande) 
                </option>
                <option
                  value="[object Object]"
                >
                  Column B (stigande) 
                </option>
                <option
                  value="[object Object]"
                >
                  Column B (fallande) 
                </option>
                
                
              </select>
              <!--v-if-->
              <f-icon-stub
                class="select-field__icon"
                library="f"
                name="arrow-down"
              />
            </div>
          </div>
          
        </div>
        
      </div>
      
    </div>
    
  </div>
  
  
</div>
`;
